consistent-sampling/src/main/java/io/opentelemetry/contrib/sampler/consistent2/ConsistentRateLimitingSampler.java

```diff
@@ -0,0 +1,157 @@
+ *
+ * <p>{@code ..., X(n-2), X(n-1), X(n)}
+ *
+ * <p>at irregular times
+ *
+ * <p>{@code ..., t(n-2), t(n-1), t(n)}
+ *
+ * <p>as
+ *
+ * <p>{@code E(X(n)) := A(n) * V(n)}.
+ *
+ * <p>{@code A(n)} and {@code V(n)} are computed recursively using Eq. 5 and Eq. 6 given by
+ *
+ * <p>{@code A(n) = b(n) * A(n-1) + X(n)} and {@code V(n) = V(n-1) / (b(n) + V(n-1))}
+ *
+ * <p>where
+ *
+ * <p>{@code b(n) := (1 - a)^(t(n) - t(n-1)) = exp((t(n) - t(n-1)) * ln(1 - a))}.
+ *
+ * <p>Introducing
+ *
+ * <p>{@code C(n) := 1 / V(n)}
+ *
+ * <p>the recursion can be rewritten as
+ *
+ * <p>{@code A(n) = b(n) * A(n-1) + X(n)} and {@code C(n) = b(n) * C(n-1) + 1}.
+ *
+ * <p>
+ *
+ * <p>Since we want to estimate the average waiting time, our data is given by
+ *
+ * <p>{@code X(n) := t(n) - t(n-1)}.
+ *
+ * <p>
+ *
+ * <p>The following correspondence is used for the implementation:
+ *
+ * <ul>
+ *   <li>{@code effectiveWindowNanos} corresponds to {@code A(n)}
+ *   <li>{@code effectiveWindowCount} corresponds to {@code C(n)}
+ *   <li>{@code decayFactor} corresponds to {@code b(n)}
+ *   <li>{@code adaptationTimeSeconds} corresponds to {@code -1 / ln(1 - a)}
+ * </ul>
+ */
+final class ConsistentRateLimitingSampler extends ConsistentSampler {
+
+  @Immutable
+  private static final class State {
+    private final double effectiveWindowCount;
+    private final double effectiveWindowNanos;
+    private final long lastNanoTime;
+
+    public State(double effectiveWindowCount, double effectiveWindowNanos, long lastNanoTime) {
+      this.effectiveWindowCount = effectiveWindowCount;
+      this.effectiveWindowNanos = effectiveWindowNanos;
+      this.lastNanoTime = lastNanoTime;
+    }
+  }
+
+  private final String description;
+  private final LongSupplier nanoTimeSupplier;
+  private final double inverseAdaptationTimeNanos;
+  private final double targetSpansPerNanosecondLimit;
+  private final AtomicReference<State> state;
+
+  /**
+   * Constructor.
+   *
+   * @param targetSpansPerSecondLimit the desired spans per second limit
+   * @param adaptationTimeSeconds the typical time to adapt to a new load (time constant used for
+   *     exponential smoothing)
+   * @param randomValueGenerator the function to use for generating the r-value
+   * @param nanoTimeSupplier a supplier for the current nano time
+   */
+  ConsistentRateLimitingSampler(
+      double targetSpansPerSecondLimit,
+      double adaptationTimeSeconds,
+      RandomValueGenerator randomValueGenerator,
+      LongSupplier nanoTimeSupplier) {
+    super(randomValueGenerator);
+
+    if (targetSpansPerSecondLimit < 0.0) {
+      throw new IllegalArgumentException("Limit for sampled spans per second must be nonnegative!");
+    }
+    if (adaptationTimeSeconds < 0.0) {
+      throw new IllegalArgumentException("Adaptation rate must be nonnegative!");
+    }
+    this.description =
+        String.format(
+            "ConsistentRateLimitingSampler{%.6f, %.6f}",
```

Comment:
To be more informative, it can say something like "target=" and "adaptation=" before the raw numerical values. The descriptions for other Samplers seem to be more verbose.

Reply:
resolved by a0c1adf165c520856d5208d5556aa15960cd9acd